File: src/main/java/com/kingmang/lazurite/parser/AST/Expressions/FunctionalExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.exceptions.VariableDoesNotExistsException;
import com.kingmang.lazurite.core.*;
import com.kingmang.lazurite.libraries.Keyword;
import com.kingmang.lazurite.parser.AST.Statements.Statement;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.Types.LzrFunction;
import com.kingmang.lazurite.runtime.LzrValue;
import com.kingmang.lazurite.runtime.Variables;


import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;


public final class FunctionalExpression extends InterruptableNode implements Expression, Statement {
    public final Expression functionExpr;
    public final List<Expression> arguments;
    
    public FunctionalExpression(Expression functionExpr) {
        this.functionExpr = functionExpr;
        arguments = new ArrayList<>();
    }
    
    public void addArgument(Expression arg) {
        arguments.add(arg);
    }
    
    @Override
    public void execute() {
        eval();
    }
    
    @Override
    public LzrValue eval() {
        super.interruptionCheck();
        final int size = arguments.size();
        final LzrValue[] values = new LzrValue[size];
        for (int i = 0; i < size; i++) {
            values[i] = arguments.get(i).eval();
        }
        final Function f = consumeFunction(functionExpr);
        CallStack.enter(functionExpr.toString(), f);
        try {
            final LzrValue result = f.execute(values);
            CallStack.exit();
            return result;
        } catch (LZRException ex) {
            throw new RuntimeException(ex.getMessage() + " in function " + functionExpr, ex);
        }
    }
    
    private Function consumeFunction(Expression expr) {
        try {
            final LzrValue value = expr.eval();
            if (value.type() == Types.FUNCTION) {
                return ((LzrFunction) value).getValue();
            }
            return getFunction(value.asString());
        } catch (VariableDoesNotExistsException ex) {
            return getFunction(ex.getVariable());
        }
    }
    
    private Function getFunction(String key) {
        if (Keyword.isExists(key)) {
            return Keyword.get(key);
        }
        if (Variables.isExists(key)) {
            final LzrValue variable = Variables.get(key);
            if (variable.type() == Types.FUNCTION) {
                return ((LzrFunction)variable).getValue();
            }
        }
        throw new LZRException("UnknownFunctionException ","Unknown function "+ key);
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        if (functionExpr instanceof ValueExpression && ((ValueExpression)functionExpr).value.type() == Types.STRING) {
            sb.append(((ValueExpression)functionExpr).value.asString()).append('(');
        } else {
            sb.append(functionExpr).append('(');
        }
        final Iterator<Expression> it = arguments.iterator();
        if (it.hasNext()) {
            sb.append(it.next());
            while (it.hasNext()) {
                sb.append(", ").append(it.next());
            }
        }
        sb.append(')');
        return sb.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/FunctionalExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.core.CallStack
import com.kingmang.lazurite.core.ClassDeclarations
import com.kingmang.lazurite.core.Function
import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.exceptions.IFileInfoProvider
import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.exceptions.LzrTracedException
import com.kingmang.lazurite.exceptions.LzrTracedException.TraceInfo
import com.kingmang.lazurite.exceptions.VariableDoesNotExistsException
import com.kingmang.lazurite.libraries.Keyword
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.Variables.get
import com.kingmang.lazurite.runtime.Variables.isExists
import com.kingmang.lazurite.runtime.values.LzrFunction
import com.kingmang.lazurite.runtime.values.LzrValue

data class FunctionalExpression(val functionExpr: Expression, val arguments: MutableList<Expression>) : InterruptableNode(), Expression, Statement {
    constructor(functionExpr: Expression) : this(functionExpr, ArrayList())

    fun addArgument(arg: Expression) {
        this.arguments.add(arg)
    }

    override fun execute() {
        eval()
    }

    override fun eval(): LzrValue {
        super.interruptionCheck()
        val values = this.arguments.map { it.eval() }.toTypedArray()
        val f = consumeFunction(this.functionExpr)
        if (this.functionExpr is IFileInfoProvider)
            CallStack.enter(this.functionExpr.toString(), f, this.functionExpr.file)
        else CallStack.enter(this.functionExpr.toString(), f, null)
        try {
            val result = f.execute(*values)
            CallStack.exit()
            return result
        } catch (ex: LzrTracedException) {
            throw ex
        } catch (ex: LzrException) {
            val info =
                if (this.functionExpr is IFileInfoProvider)
                    TraceInfo(this.functionExpr.file, CallStack.getCalls())
                else TraceInfo(null, CallStack.getCalls())
            CallStack.exit()
            throw LzrTracedException(ex.type, ex.message!!, info)
        }
    }

    private fun consumeFunction(expr: Expression): Function {
        try {
            val value = expr.eval()
            if (value.type() == Types.FUNCTION)
                return (value as LzrFunction).value
            return getFunction(value.asString())
        } catch (ex: VariableDoesNotExistsException) {
            return getFunction(ex.variable)
        }
    }

    private fun getFunction(key: String): Function {
        if (Keyword.isExists(key))
            return Keyword.get(key)
        if (isExists(key)) {
            val variable = get(key)
            if (variable!!.type() == Types.FUNCTION) {
                return (variable as LzrFunction?)!!.value
            }
        }

        if (ClassDeclarations[key] != null) {
            throw LzrException("UnknownFunctionException", "Unknown function $key. You trying to create an instance of class $key? Use the keyword 'new'")
        } else {
            throw LzrException("UnknownFunctionException", "Unknown function $key")
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R =
        visitor.visit(this, t)

    override fun toString(): String {
        val sb = StringBuilder()
        if (functionExpr is ValueExpression && functionExpr.value.type() == Types.STRING)
            sb.append(functionExpr.value.asString()).append('(')
        else sb.append(functionExpr).append('(')
        val it: Iterator<Expression> = arguments.iterator()
        if (it.hasNext()) {
            sb.append(it.next())
            while (it.hasNext()) {
                sb.append(", ").append(it.next())
            }
        }
        sb.append(')')
        return sb.toString()
    }
}
